xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2019, Digi International Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES 
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR 
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES 
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN 
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF 
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+package com.digi.xbee.api.android.connection.android;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.digi.xbee.api.utils.HexUtils;
+
+import android.hardware.usb.UsbDeviceConnection;
+import android.hardware.usb.UsbEndpoint;
+
+/**
+ * This class acts as a wrapper to write data to the USB Interface in Android
+ * behaving like an {@code OutputStream} class.
+ */
+public class AndroidUSBOutputStream extends OutputStream {
+
+	// Constants.
+	private static final int WRITE_TIMEOUT = 2000;
+	
+	// Variables.
+	private UsbDeviceConnection usbConnection;
+
+	private UsbEndpoint sendEndPoint;
+	
+	private Logger logger;
+
+	/**
+	 * Class constructor. Instantiates a new {@code AndroidUSBOutputStream}
+	 * object with the given parameters.
+	 * 
+	 * @param writeEndpoint The USB end point to use to write data to.
+	 * @param connection The USB connection to use to write data to.
+	 * 
+	 * @see UsbDeviceConnection
+	 * @see UsbEndpoint
+	 */
+	public AndroidUSBOutputStream(UsbEndpoint writeEndpoint, UsbDeviceConnection connection) {
+		this.usbConnection = connection;
+		this.sendEndPoint = writeEndpoint;
+		this.logger = LoggerFactory.getLogger(AndroidUSBOutputStream.class);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.io.OutputStream#write(int)
+	 */
+	@Override
+	public void write(int oneByte) throws IOException {
+		write(new byte[] {(byte)oneByte});
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.io.OutputStream#write(byte[])
+	 */
+	@Override
+	public void write(byte[] buffer) throws IOException {
+		write(buffer, 0, buffer.length);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.io.OutputStream#write(byte[], int, int)
+	 */
+	@Override
+	public void write(byte[] buffer, int offset, int count) throws IOException {
+		final byte[] finalData = new byte[count + offset];
```

Comment:
Do we really want to launch a new thread for every write call? Seems more logical to me to have a long-running thread onto which we schedule the writes.

Reply:
To be totally honest I didn't took a deep look to the code as I was just moving classes from one lib to another. Anyway you're right, a single thread taking data from a queue is a much better approach. I'll change it =)

Reply:
This is the new implementation so far:
```java
private LinkedBlockingQueue<byte[]> writeQueue;

public AndroidUSBOutputStream(UsbEndpoint writeEndpoint, UsbDeviceConnection connection) {
	[...]

	writeQueue = new LinkedBlockingQueue<>(512);
	DataWriter dataWriter = new DataWriter();
	dataWriter.start();
}

public void write(byte[] buffer, int offset, int count) {
	final byte[] finalData = new byte[count + offset];
	System.arraycopy(buffer, offset, finalData, 0, count);
	try {
		writeQueue.add(finalData);
	} catch (IllegalStateException  e) {
		logger.error("Could not add data, write queue is full: " + e.getMessage(), e);
	}
}

/**
 * Internal class used to write data coming from a queue.
 */
class DataWriter extends Thread {
	@Override
	public void run() {
		while (streamOpen) {
			try {
				byte[] dataToWrite = writeQueue.poll(100, TimeUnit.MILLISECONDS);
				if (dataToWrite == null)
					continue;
				usbConnection.bulkTransfer(sendEndPoint, dataToWrite, dataToWrite.length, WRITE_TIMEOUT);
				logger.debug("Message sent: " + HexUtils.byteArrayToHexString(dataToWrite));
			} catch (InterruptedException e) {
				logger.error("Interrupted while getting data from the write queue: " + e.getMessage(), e);
			}
		}
	}
}

@Override
public void close() throws IOException {
	// Stop the data writer.
	streamOpen = false;
	super.close();
}
```